### PR TITLE
Pokeball

### PR DESCRIPTION
**Pull request by ghoulslash**
decompile asm/pokeball.s & data/pokeball.s into src/pokeball.c

### CODE REVIEW

File: ld_script.txt [hunk 1]
```diff
@@ -83,7 +83,7 @@ SECTIONS {
         src/daycare.o(.text);
         asm/battle_interface.o(.text);
         asm/smokescreen.o(.text);
-        asm/pokeball.o(.text);
+		src/pokeball.o(.text);
```
Comment (jiangzhengwenjz) [inline ld_script.txt]: tab -> space

File: src/pokeball.c [hunk 2]
```diff
@@ -0,0 +1,1543 @@
...
+    gBattleSpritesDataPtr->healthBoxesData[gActiveBattler].ballAnimActive = 1;
+    taskId = CreateTask(Task_DoPokeballSendOutAnim, 5);
+    gTasks[taskId].tPan = pan;
+    gTasks[taskId].tThrowId = kindOfThrow;
+    gTasks[taskId].tBattler = gActiveBattler;
+    return 0;
+}
+
+#ifdef NONMATCHING
+//gender is an unused variable that still manages to show up in the assembly, can't get temporary variables to make it work.
```
Comment (jiangzhengwenjz) [inline src/pokeball.c]: Compiler explorer
Is it in emerald?

File: src/pokeball.c [hunk 3]
```diff
@@ -0,0 +1,1543 @@
...
+        }
+    }
+    else
+    {
+        if (TranslateAnimHorizontalArc(sprite))
+        {
+            sprite->pos1.x += sprite->pos2.x;
+            sprite->pos1.y += sprite->pos2.y;
+            sprite->pos2.y = 0;
+            sprite->pos2.x = 0;
```
Comment (jiangzhengwenjz) [inline src/pokeball.c]: Not a big issue, but it's more like `sprite->pos2.x = sprite->pos2.y = 0;`

File: src/pokeball.c [hunk 4]
```diff
@@ -0,0 +1,1543 @@
...
+{
+    if (TranslateAnimHorizontalArc(sprite))
+    {
+        u16 ballId;
+        u8 taskId = sprite->oam.affineParam;
+        u8 opponentBattler = gTasks[taskId].tOpponentBattler;
+        u8 noOfShakes = gTasks[taskId].tThrowId;
+
+        StartSpriteAnim(sprite, 1);
+        sprite->affineAnimPaused = 1;
```
Comment (jiangzhengwenjz) [inline src/pokeball.c]: `TRUE`

File: src/pokeball.c (hunk 2 above)
Comment (ghoulslash) [reply src/pokeball.c]: https://cexplore.karathan.at/z/hGN7WV

It is not a direct port from emerald.

File: src/pokeball.c [hunk 5]
```diff
@@ -1216,8 +1216,7 @@ static void SpriteCB_PlayerMonSendOut_2(struct Sprite *sprite)
         {
             sprite->pos1.x += sprite->pos2.x;
             sprite->pos1.y += sprite->pos2.y;
-            sprite->pos2.y = 0;
-            sprite->pos2.x = 0;
+            sprite->pos2.y = sprite->pos2.x = 0;
```
Comment (jiangzhengwenjz) [inline src/pokeball.c]: Incorrect. The second assign will be executed first

File: src/pokeball.c (hunk 2 above)
Comment (jiangzhengwenjz) [reply src/pokeball.c]: https://cexplore.karathan.at/z/p6UZDJ check this

File: src/pokeball.c [hunk 6]
```diff
@@ -105,9 +105,9 @@ const struct CompressedSpritePalette gBallSpritePalettes[POKEBALL_COUNT] =
 static const struct OamData sBallOamData =
 {
     .y = 0,
-    .affineMode = 3,
-    .objMode = 0,
-    .mosaic = 0,
+    .affineMode = ST_OAM_AFFINE_DOUBLE,
+    .objMode = ST_OAM_OBJ_NORMAL,
+    .mosaic = ST_OAM_4BPP,
```
Comment (jiangzhengwenjz) [inline src/pokeball.c]: `FALSE`

File: src/pokeball.c [hunk 7]
```diff
@@ -105,9 +105,9 @@ const struct CompressedSpritePalette gBallSpritePalettes[POKEBALL_COUNT] =
...
     .bpp = 0,
```
Comment (jiangzhengwenjz) [inline src/pokeball.c]: `ST_OAM_4BPP`

File: src/pokeball.c [hunk 8]
```diff
@@ -0,0 +1,1275 @@
...
+    },
+};
+
+// Functions
+u8 DoPokeballSendOutAnimation(s16 pan, u8 kindOfThrow)
+{
+    u8 taskId;
+    
+    gDoingBattleAnim = TRUE;
+    gBattleSpritesDataPtr->healthBoxesData[gActiveBattler].ballAnimActive = 1;
```
Comment (jiangzhengwenjz) [inline src/pokeball.c]: `TRUE`

File: src/pokeball.c [hunk 9]
```diff
@@ -0,0 +1,1275 @@
...
+
+        gSprites[gBattlerSpriteIds[battlerId]].pos2.y = 0;
+        gDoingBattleAnim = FALSE;
+        gBattleSpritesDataPtr->healthBoxesData[battlerId].ballAnimActive = 0;
+        FreeSpriteOamMatrix(sprite);
+        DestroySprite(sprite);
+
+        for (doneBattlers = 0, i = 0; i < MAX_BATTLERS_COUNT; i++)
+        {
+            if (gBattleSpritesDataPtr->healthBoxesData[i].ballAnimActive == 0)
```
Comment (jiangzhengwenjz) [inline src/pokeball.c]: `if (!gBattleSpritesDataPtr->healthBoxesData[i].ballAnimActive)`

File: src/pokeball.c [hunk 10]
```diff
@@ -0,0 +1,1275 @@
...
+        gSprites[gBattlerSpriteIds[battlerId]].data[1] -= 288;
+        gSprites[gBattlerSpriteIds[battlerId]].pos2.y = gSprites[gBattlerSpriteIds[battlerId]].data[1] >> 8;
+    }
+    if (sprite->animEnded && affineAnimEnded)
+    {
+        s32 i, doneBattlers;
+
+        gSprites[gBattlerSpriteIds[battlerId]].pos2.y = 0;
+        gDoingBattleAnim = FALSE;
+        gBattleSpritesDataPtr->healthBoxesData[battlerId].ballAnimActive = 0;
```
Comment (jiangzhengwenjz) [inline src/pokeball.c]: `FALSE`

File: src/pokeball.c [hunk 11]
```diff
@@ -0,0 +1,1275 @@
...
+        m4aMPlayAllStop();
+        PlaySE(MUS_FAN6);
+    }
+    else if (sprite->data[4] == 315)
+    {
+        FreeOamMatrix(gSprites[gBattlerSpriteIds[sprite->sBattler]].oam.matrixNum);
+        DestroySprite(&gSprites[gBattlerSpriteIds[sprite->sBattler]]);
+        DestroySpriteAndFreeResources(sprite);
+        if (gMain.inBattle)
+            gBattleSpritesDataPtr->healthBoxesData[battlerId].ballAnimActive = 0;
```
Comment (jiangzhengwenjz) [inline src/pokeball.c]: `FALSE`

File: src/pokeball.c [hunk 12]
```diff
@@ -0,0 +1,1275 @@
...
+        gSprites[r1].pos2.y = 0;
+        DestroySprite(sprite);
+    }
+}
+
+void LoadBallGfx(u8 ballId)
+{
+    u16 var;
+
+    if (GetSpriteTileStartByTag(gBallSpriteSheets[ballId].tag) == 0xFFFF)
```
Comment (jiangzhengwenjz) [inline src/pokeball.c]: `SPRITE_INVALID_TAG`

**Pull request merged**